Source/qol/stash.cpp

```diff
@@ -494,53 +449,64 @@ bool UseStashItem(uint16_t c)
 
 	UseItem(MyPlayerId, item->_iMiscId, item->_iSpell);
 
-	if (Stash.stashList[c]._iMiscId == IMISC_MAPOFDOOM)
+	if (Stash.stashList[itemIndex]._iMiscId == IMISC_MAPOFDOOM)
 		return true;
-	if (Stash.stashList[c]._iMiscId == IMISC_NOTE) {
+	if (Stash.stashList[itemIndex]._iMiscId == IMISC_NOTE) {
 		InitQTextMsg(TEXT_BOOK9);
 		CloseInventory();
 		return true;
 	}
-	Stash.RemoveStashItem(c);
+	Stash.RemoveStashItem(item, itemPosition, itemIndex);
 
 	return true;
 }
 
-void StashStruct::RemoveStashItem(StashStruct::StashCell iv)
+void StashStruct::RemoveStashItem(Item *item, Point itemPosition, StashCell itemIndex)
 {
-	// Iterate through stashGrid and remove every reference to item
-	for (auto &row : Stash.GetCurrentGrid()) {
-		for (StashStruct::StashCell &itemId : row) {
-			if (itemId - 1 == iv) {
-				itemId = 0;
-			}
-		}
-	}
-
-	if (stashList.empty()) {
+	if (item == nullptr) {
 		return;
 	}
 
-	// If the item at the end of stash array isn't the one we removed, we need to swap its position in the array with the removed item
+	// get item's size to figure slot changes
+	const Size itemSize = GetInventorySize(*item);
+
+	// reset claimed slot IDs to zero
+	UpdateStashGrid(Stash.GetPage(), itemPosition, itemSize);
+
+	// not the last item, swap with last item to keep IDs contiguous

```

Comment:
this comment should move into the if statement

Reply:
Done, thank you!